### PR TITLE
fix dma unregister callback during region removal

### PR DESCRIPTION
**Pull request by jlevon**

### CODE REVIEW

File: include/libvfio-user.h [hunk 1]
```diff
@@ -655,7 +655,9 @@ vfu_unmap_sg(vfu_ctx_t *vfu_ctx, const dma_sg_t *sg,
              struct iovec *iov, int cnt);
 
 /**
- * Read from the dma region exposed by the client.
+ * Read from the dma region exposed by the client. This can be used as an
+ * alternative to vfu_map_sg(), if the region is not directly mappable, or DMA
+ * callbacks have not been provided.
```
Comment (swapnili) [inline include/libvfio-user.h:660]: Should we be more clear and call it as "DMA notification callbacks"

File: include/libvfio-user.h [hunk 2]
```diff
@@ -667,7 +669,9 @@ int
 vfu_dma_read(vfu_ctx_t *vfu_ctx, dma_sg_t *sg, void *data);
 
 /**
- * Write to the dma region exposed by the client.
+ * Write to the dma region exposed by the client.  This can be used as an
+ * alternative to vfu_map_sg(), if the region is not directly mappable, or DMA
+ * callbacks have not been provided.
```
Comment (swapnili) [inline include/libvfio-user.h:674]: ditto "DMA notification callbacks"

File: lib/dma.c [hunk 3]
```diff
@@ -178,21 +178,32 @@ MOCK_DEFINE(dma_controller_remove_region)(dma_controller_t *dma,
 }
 
 void
-dma_controller_remove_regions(dma_controller_t *dma)
+dma_controller_remove_regions(dma_controller_t *dma,
```
Comment (swapnili) [inline lib/dma.c:181]: I think we should rename this func to dma_controller_remove_all_regions()
This name really confuses with dma_controller_remove_region()

File: lib/dma.c [hunk 4]
```diff
@@ -207,11 +218,6 @@ dma_controller_remove_regions(dma_controller_t *dma)
 void
 dma_controller_destroy(dma_controller_t *dma)
 {
-    if (dma == NULL) {
```
Comment (swapnili) [inline lib/dma.c:210]: Why following snippet is removed?

Comment (jlevon) [reply lib/dma.c:210]: no need to check NULL when we only call free()

File: lib/dma.c (hunk 3 above)
Comment (jlevon) [reply lib/dma.c:181]: sure, done.

File: include/libvfio-user.h (hunk 2 above)
Comment (jlevon) [reply include/libvfio-user.h:674]: done

File: include/libvfio-user.h (hunk 1 above)
Comment (jlevon) [reply include/libvfio-user.h:660]: done

File: lib/dma.c [hunk 5]
```diff
@@ -207,11 +218,6 @@ dma_controller_remove_regions(dma_controller_t *dma)
...
-        return;
-    }
-
-    dma_controller_remove_regions(dma);
```
Comment (swapnili) [inline lib/dma.c:214]: Sorry I think previous comment was not clear. I mean to say that why call to dma_controller_remove_regions() is been removed.

**Review by swapnili — APPROVED** [review]

**Pull request merged**